src/cargo-fmt/main.rs

```diff
@@ -748,4 +861,197 @@ mod cargo_fmt_tests {
             );
         }
     }
+
+    mod get_nested_integration_test_files_tests {
+        use super::*;
+
+        #[test]
+        fn returns_no_files_if_root_not_dir() {
+            let target_dir = PathBuf::from("tests/nested-test-files/no-test-dir/Cargo.toml");
+            assert_eq!(
+                Vec::new() as Vec<PathBuf>,
+                get_nested_integration_test_files(&target_dir, &target_dir),
+            )
+        }
+
+        #[test]
+        fn returns_no_files_if_tests_has_no_nested_files() {
+            let target_dir = Path::new("tests/nested-test-files/only-root-level-tests/tests");
+            assert_eq!(
+                Vec::new() as Vec<PathBuf>,
+                get_nested_integration_test_files(&target_dir, &target_dir),
+            )
+        }
+
+        #[test]
+        fn returns_nested_files() {
+            let target_dir = Path::new("tests/nested-test-files/root-and-nested-tests/tests");
+            let exp_baz = PathBuf::from(
+                "tests/nested-test-files/root-and-nested-tests/tests/nested/deeply-nested/baz.rs",
+            );
+            let exp_foo_bar = PathBuf::from(
+                "tests/nested-test-files/root-and-nested-tests/tests/nested/foo_bar.rs",
+            );
+            let exp_other = PathBuf::from(
+                "tests/nested-test-files/root-and-nested-tests/tests/nested/other.rs",
+            );
+            assert_eq!(
+                vec![exp_baz, exp_foo_bar, exp_other],
+                get_nested_integration_test_files(&target_dir, &target_dir),
+            )
+        }
+    }
+
+    mod get_targets_tests {
+        use super::*;
+
+        fn create_stub_target(
+            src_path: &str,
+            test_files: Vec<PathBuf>,
+            kind: &str,
+            edition: &str,
+        ) -> Target {
+            let path = PathBuf::from(src_path);
+            let canonicalized = fs::canonicalize(&path).unwrap_or(path);
+            Target {
+                path: canonicalized,
+                kind: String::from(kind),
+                edition: String::from(edition),
+                nested_int_test_files: test_files,
+            }
+        }
+
```

Comment:
Running the tests directly against the `get_targets` [function](https://github.com/rust-lang/rustfmt/blob/a15e97f1e92b3621ab8c2f1585f27e8de583f74b/src/cargo-fmt/main.rs#L310) using the same sample projects under the `tests` directory. I wasn't originally planning on testing the new behavior this way, but it turns out to be rather tricky to instantiate some of the structs from `cargo_metadata` (no constructors, private struct fields, etc.) that would have been needed to test the functions that work with `cargo_metadata` structs directly (for example the `add_targets` function). 

I think this actually turns out to be a good thing though as these tests help validate the target determination process more holistically.